### PR TITLE
AP_Periph for UAVCAN to various RCOUT translations

### PR DESCRIPTION
**Pull request by bugobliterator**
How to use the latest version:

Actuator UAVCAN msgs id 1 - 16 will get mapped to OUTn_FUNCTION = 51-66 (rcin1 - rc16)
Example: Actuator id 2 will always map to FUNCTION = 52 (RCin2)

ESC RawArray UAVCAN msgs index 0 - 11 get mapped to:
OUTn_FUNCTION = 33 - 40, and 82 - 85 (motor1 motor 8, motor 9 - motor 12)

So, set up whatever "OUTn" to whatever ESC/Actuator you want. You can even do duplicates, such as set multiple servos to 51 and they will all move when the input changes, but also you can set different min/max/trims so you actually get different outputs all scaled to that same input. Fancy!

We'll need to figure out a way to document this so people know how to set the OUTn_FUNCTION values correctly.

### CODE REVIEW

**Comment by magicrub** [comment]
I'm in the process of testing this.

**Comment by davidbitton** [comment]
What parameters need to be set in the node as well as the flight controller to get this working? Thanks.

**Comment by magicrub** [comment]
I'm having trouble testing this but I think it's a hardware bootloader problem on my end. I'll test it next week when I get a programmer+cable

**Comment by bugobliterator** [comment]
@davidbitton there is a README.md under AP_Periph, which has those details.

**Comment by davidbitton** [comment]
> @davidbitton there is a README.md under AP_Periph, which has those details.

The readme doesn't contain Ardupilot settings or CAN node settings. All it speaks to is building and flashing.

**Comment by magicrub** [comment]
I'm testing this. There are problems keeping servo from working but I'm working through them..

**Comment by magicrub** [comment]
@davidbitton @bugobliterator Ok, I've done a crazy renaming job on this so I'm sorry for the messy commit. The big problem I was having was:
- "PWM 1" actually meant 2 (it was zero indexed)
- Cube Black/Orange-periph do not use IO_MPU so "PWM 1" is Cube AUX 1

I've also reworked some of the logic so all the settings can change at runtime

**Comment by ukmook** [comment]
I have tested this and happy to report that servos are working as expected. 
PWM1 = pin1, PWM2 = pin2 and so on. 
I am able to change servo option and the node updates I.e. from aileron to elevator.

**Comment by magicrub** [comment]
@bugobliterator I've added the EuDev call tag and removed WIP. I pounded on it pretty good and there's multiple people using it now. I think it's ready to merge.

That said, I think there's some feature-creep that would be nice but we can add it on a follow-up PR.
- PWM_MIN/MAX for each channel
- allow both ESC and servos at same time (change ACT_TYPE to a bitmask where 0 is srv and 1 is ESC?)
- rework CHAN_LOW/HIGH to be more flexible (wish list item, it works fine as-is for now)

**Review by magicrub — APPROVED** [review]

File: Tools/AP_Periph/rcout_translator.cpp [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        hal.rcout->write(i - rcout.chan_start - 1, output_pwm);
+    }
+#if HAL_NUM_CAN_IFACES > 1
+    if (rcout_params.can_out == 1) {
+        switch (rcout.protocol) {
+            case RCOUTTranslator_Params::RCOUT_KDECAN: {
+                if (rcout_params.kdecan == nullptr) {
+                    return;
+                }
+                rcout_params.kdecan->lock_rcout();
```
Comment (tridge) [inline Tools/AP_Periph/rcout_translator.cpp:178]: use WITH_SEMAPHORE()

**Review by tridge — CHANGES_REQUESTED** [review]
remove KDECAN until it is working on real hw
or wait for KDECAN

File: Tools/AP_Periph/rcout_translator.cpp (hunk 1 above)
Comment (magicrub) [reply Tools/AP_Periph/rcout_translator.cpp]: removed KDECAN

File: Tools/AP_Periph/can.cpp [hunk 2]
```diff
@@ -426,6 +428,7 @@ static void handle_allocation_response(CanardInstance* ins, CanardRxTransfer* tr
 /*
   fix value of a float for canard float16 format
  */
+#pragma GCC diagnostic ignored "-Wunused-function"
```
[the remark this answers is not in the document]
Comment (magicrub) [reply Tools/AP_Periph/can.cpp:442]: added push/pop

**Comment by magicrub** [comment]
I removed KDECAN until we can prove it works. I also renamed param RCOUT_PROTOCOL to RCOUT_PWM_TYPE and RCOUT_ACT_TYPE to RCOUT_OUT_TYPE because that makes WAY more sense.

**Comment by magicrub** [comment]
I've removed KDECAN, cleaned it up a bit ad rebased. @ukmook could you do me a favor and give this a quick test again?

**Comment by davidbitton** [comment]
@magicrub is the code now in master?

**Comment by magicrub** [comment]
@davidaroyer no, this pull-request is still open. It will be in master when the status changes to "merged". Are you able to test this branch?

**Comment by davidbitton** [comment]
@magicrub i was going to try it. so, you're making changes against @bugobliterator's repo? that's where I should pull from?

**Comment by magicrub** [comment]
@davidbitton yes. The branch on this PR.

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
this is looking really nice now. I'd just like to see some testing on existing hw, for example, I'd like to test on the CUAV GPS which does the fancy LEDs to make sure that is still OK

**Comment by magicrub** [comment]
@tridge I have been testing on Cube Orange and Cube Black and @ukmook has tested on his own custom PA_Periph hardware. There's been activity on the #canbus and then onto #ap_periph discord channels.
https://discord.com/channels/674039678562861068/711798608147644416/776873738443554876

**Comment by ukmook** [comment]
I have a can gps with fancy leds. :) notify ncp5263, I’ll test as soon as I can, which should be tomorrow evening.

**Comment by magicrub** [comment]
This needs a rebase to fix a merge conflict from the recent battery monitor merge

**Comment by magicrub** [comment]
I propose some new params where n is for each available pwm output:
RCOUT_OUTn_MIN
RCOUT_OUTn_MAX
RCOUT_OUTn_FN (srv or esc)
RCOUT_OUTn_TYPE (pwm, dshot..)
RCOUT_OUTn_MAP (map this UAVCAN msg index to this channel n)

What about frequency?

I'd make a 5 object param that would have an instance per available PWM() value in the hwdef. Similar to SERVOn_ on the vehicle code.

**Comment by ukmook** [comment]
I have tested this pr again after rebase on an f413 node similar to cuav-gps (f412), rcout is working with servos along with batt-mon, neo-pixel, baro, mag and gps. Flight controller was H743 type.

**Comment by davidbitton** [comment]
> RCOUT_OUTn_TYPE (pwm, dshot..)

Is it Type or Protocol? Why not RCOUT_OUTn_FUNCTION to match SERVOn_FUNCTION, or do you plan on following the SERVOn_FUNCTION setting?

**Comment by magicrub** [comment]
This morning I attempted to bring in the whole servo library. I need more time.

File: Tools/AP_Periph/wscript [hunk 3]
```diff
@@ -37,6 +37,10 @@ def build(bld):
         'AP_RangeFinder',
         'AP_ROMFS',
         'AP_MSP',
+        'SRV_Channel',
+        'AP_Volz_Protocol',
```
Comment (tridge) [inline Tools/AP_Periph/wscript:41]: we could ifdef for AP_Periph to avoid this

**Review by tridge — CHANGES_REQUESTED** [review]

File: Tools/AP_Periph/wscript (hunk 3 above)
Comment (magicrub) [reply Tools/AP_Periph/wscript]: removed

**Comment by magicrub** [comment]
How to use the latest version:

Actuator UAVCAN msgs id 1 - 16 will get mapped to OUTn_FUNCTION = 51(rcin1) ... 66(rcin16)

ESC RawArray UAVCAN msgs index 0 - 11 get mapped to:
OUTn_FUNCTION = 33(motor1) ... 40(motor8), 82(motor9) ... 85(motor12)

So, set up whatever "OUTn" to whatever ESC/Actuator you want. You can even do duplicates, such as set multiple servos to 51 and they will all move when the input changes, but also you can set different min/max/trims so you actually get different outputs all scaled to that same input. Fancy!

We'll need to figure out a way to document this so people know how to set the OUTn_FUNCTION values correctly

**Comment by magicrub** [comment]
Ok, this has FINALLY been fixed and is ready for prime time.

Turns out libcanard has a nasty bug in it's packet decoder un-serializer helper function when packets contain multiple arrays. This bug appears to only be in the reduced program-size verison of uavcan called libcanard and does not effect ArduPilot vehicle code.

@ukmook and/or @davidbitton could you guys give this a spin one more time?

**Comment by ukmook** [comment]
Tested again, all servos now working correctly, along with mag, baro, current sensing and neo pixel all working on same node.

**Comment by tridge** [comment]
I think we shoud have a SAFETY_DEFAULT parameter, equivalent to BRD_SAFETY_ENABLE
so value 0 means default off, 1 means default on, should default to 1

**Comment by tridge** [comment]
need to stop sending pulses before reboot, then delay 1ms, to stop chopped pulses

**Comment by tridge** [comment]
please check that f103-GPS fw size doesn't change with this PR

**Comment by magicrub** [comment]
@tridge the reboot suggestion has been added. Does the board safety need to be included in this PR or can we get that in the next? I still need to check the f1 file size per your request

**Comment by magicrub** [comment]
f103-GPS grows by 172 bytes with this PR. That is likely due to the improved reboot behavior where we set:
```hal.rcout->force_safety_on();``` (useless call if RC Out is not enabled)
```AP_Param::flush();``` (useful for all targets)
```hal.scheduler->register_delay_callback(nullptr, 5);``` (perhaps we should remove this from AP_Periph?)

**Comment by magicrub** [comment]
See commit https://github.com/ArduPilot/ardupilot/pull/15694/commits/6ffc81a5e31a9a7901eb80891fab22ae1752800d for BRD_SAFTEENABLE. I think it's actually not even needed. Trims do not get set on boot like I thought, there's no servo movement until there are received commands

**Comment by tridge** [comment]
great work! can be merged when we remove the safety param, and default to safety on

**Comment by rligocki** [comment]
Hey guys, I tried to found any information if it is possible to enable DShot output on AP-Periph. Unfortunately in source code I didn't found any way how to configure output protocol. Is it already implemented? I am able to provide my skills and some time to add these feature. I need this feature to replace UC4H ESC-Actuator firmware. Thanks

**Pull request merged**